### PR TITLE
Reconcile MTQ CR according to an HCO FeatureGate

### PR DESCRIPTION
**Pull request by orenc1**
If hco's `.spec.featureGates.enableManagedTenantQuota=true`, the MTQ CustomResource will be deployed in the installation namespace. When the feature gate is changed to `false`, the CR is deleted.
The feature is in Dev Preview (DP) and it is in opt-in mode (disabled by default).

**What this PR does / why we need it**:

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-31716
```

**Release note**:

```release-note
Reconcile MTQ CR and put it behind a feature gate in HCO. Default is off (opt-in feature).
```

### CODE REVIEW

File: controllers/operands/mtq.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hcov1beta1 "github.com/kubevirt/hyperconverged-cluster-operator/api/v1beta1"
+	"github.com/kubevirt/hyperconverged-cluster-operator/controllers/common"
+	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+	mtqv1alpha1 "kubevirt.io/managed-tenant-quota-api/pkg/apis/core/v1alpha1"
```
Comment (Barakmor1) [inline controllers/operands/mtq.go:18]: Thanks for this PR please change to :

`mtqv1alpha1 "kubevirt.io/managed-tenant-quota/staging/src/kubevirt.io/managed-tenant-quota-api/pkg/apis/core/v1alpha1" `

File: go.mod [hunk 2]
```diff
@@ -36,6 +36,7 @@ require (
 	kubevirt.io/api v1.0.0
 	kubevirt.io/containerized-data-importer-api v1.57.0-rc1
 	kubevirt.io/controller-lifecycle-operator-sdk/api v0.2.4
+	kubevirt.io/managed-tenant-quota-api v1.1.0
```
Comment (Barakmor1) [inline go.mod:39]: actually we didn't create api repo yet please change to:

`kubevirt.io/managed-tenant-quota v1.1.0`

**Review by Barakmor1 — COMMENTED** [review]
this is the reason for tha failures i guess

File: tests/func-tests/mtq_test.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+				g.Expect(mtq.Status.Conditions).ShouldNot(BeEmpty())
+				return conditionsv1.IsStatusConditionTrue(mtq.Status.Conditions, conditionsv1.ConditionAvailable)
+			}).WithTimeout(5 * time.Minute).WithPolling(time.Second).ShouldNot(BeTrue())
+
+			By("check MTQ pods")
+			pods, err := cli.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(ctx, metav1.ListOptions{LabelSelector: "app.kubernetes.io/component=multi-tenant"})
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(pods.Items).Should(HaveLen(3))
+		})
+	})
```
Comment (Barakmor1) [inline tests/func-tests/mtq_test.go:85]: Hey i'm not sure, wouldn't it be a good idea to disable the feature gate in the `beforeEach` scope as well?

Nit: we can use function to make this more readable:

```
    AfterAll(func() {
        disableMTQFeatureGate(ctx, cli)
        Eventually(checkMTQResourceNotFound).WithTimeout(5 * time.Minute).WithPolling(100 * time.Millisecond).Should(BeTrue())
    })

    When("set the EnableManagedTenantQuota FG", func() {
        It("should create the MTQ CR and all the pods", func() {
            enableMTQFeatureGate(ctx, cli)
            Eventually(checkMTQConditionAvailable).WithTimeout(5 * time.Minute).WithPolling(time.Second).ShouldNot(BeTrue())
        })
    })
})

func enableMTQFeatureGate(ctx context.Context, cli kubecli.KubevirtClient) {
    patch := []byte(fmt.Sprintf(mtqFGTemplate, true))
    Eventually(tests.PatchHCO).
        WithArguments(ctx, cli, patch).
        WithTimeout(10 * time.Second).
        WithPolling(100 * time.Millisecond).
        Should(Succeed())
}

func disableMTQFeatureGate(ctx context.Context, cli kubecli.KubevirtClient) {
    patch := []byte(fmt.Sprintf(mtqFGTemplate, false))
    Eventually(tests.PatchHCO).
        WithArguments(ctx, cli, patch).
        WithTimeout(10 * time.Second).
        WithPolling(100 * time.Millisecond).
        Should(Succeed())
}

func checkMTQResourceNotFound(g Gomega) bool {
    _, err := getMTQResource(ctx, cli)
    g.Expect(err).To(HaveOccurred())
    return errors.IsNotFound(err)
}

func checkMTQConditionAvailable(g Gomega) bool {
    mtq := getMTQ(ctx, cli, g)
    g.Expect(mtq.Status.Conditions).ShouldNot(BeEmpty())
    return conditionsv1.IsStatusConditionTrue(mtq.Status.Conditions, conditionsv1.ConditionAvailable)
}
```

**Review by Barakmor1 — APPROVED** [review]
Hey,
Thanks Nahshon i mentioned couple of nits 
overall looks great and clear to me please feel free to ignore
/lgtm

File: controllers/hyperconverged/hyperconverged_controller_test.go [hunk 4]
```diff
@@ -260,6 +260,91 @@ var _ = Describe("HyperconvergedController", func() {
...
+					Status:  metav1.ConditionFalse,
+					Reason:  reconcileInit,
+					Message: "Initializing HyperConverged cluster",
+				})))
+				Expect(foundResource.Status.Conditions).To(ContainElement(commontestutils.RepresentCondition(metav1.Condition{
+					Type:    hcov1beta1.ConditionUpgradeable,
+					Status:  metav1.ConditionUnknown,
+					Reason:  reconcileInit,
+					Message: "Initializing HyperConverged cluster",
+				})))
```
[the remark this answers is not in the document]
Comment (nunnatsa) [reply controllers/hyperconverged/hyperconverged_controller_test.go:321]: Not sure we gain too much from this change. We'll still have the explicit list of conditions in the code.

**Comment by nunnatsa** [comment]
hco-e2e-kv-smoke-gcp lane passed
/override ci/prow/hco-e2e-kv-smoke-azure
hco-e2e-upgrade-operator-sdk-aws lane passed
/override ci/prow/hco-e2e-upgrade-operator-sdk-azure
okd-hco-e2e-operator-sdk-aws lane passed
/override ci/prow/okd-hco-e2e-operator-sdk-gcp

/cherry-pick release-1.10

**Pull request merged**